cellarium/ml/layers/attention.py

```diff
@@ -0,0 +1,164 @@
+        d_model:
+            Dimensionality of the embeddings and hidden states.
+        use_bias:
+            Whether to use bias in the linear transformations.
+        n_heads:
+            Number of attention heads.
+        dropout_p:
+            Dropout probability.
+        attention_logits_scale:
+            Multiplier for the attention scores.
+        attention_backend:
+            Backend for the attention computation.
+        attention_softmax_fp32:
+            Whether to use float32 for softmax computation.
+        Wqkv_initializer:
+            Initializer for the query, key, and value linear transformations.
+        Wo_initializer:
+            Initializer for the output linear transformation.
+    """
+
+    backend_map = {
+        "math": SDPBackend.MATH,
+        "flash": SDPBackend.FLASH_ATTENTION,
+        "mem_efficient": SDPBackend.EFFICIENT_ATTENTION,
+    }
+
+    def __init__(
+        self,
+        d_model: int,
+        use_bias: bool,
+        n_heads: int,
+        dropout_p: float,
+        attention_logits_scale: float,
+        attention_backend: Literal["math", "flash", "mem_efficient", "torch"],
+        attention_softmax_fp32: bool,
+        Wqkv_initializer: dict[str, Any],
+        Wo_initializer: dict[str, Any],
+    ) -> None:
+        super().__init__()
+        self.Wq = nn.Linear(d_model, d_model, bias=use_bias)
+        self.Wk = nn.Linear(d_model, d_model, bias=use_bias)
+        self.Wv = nn.Linear(d_model, d_model, bias=use_bias)
+        self.Wo = nn.Linear(d_model, d_model, bias=use_bias)
+        self.n_heads = n_heads
+        self.dropout_p = dropout_p
+        self.attention_logits_scale = attention_logits_scale
+        self.attention_backend = attention_backend
+        self.attention_softmax_fp32 = attention_softmax_fp32
+        self.Wqkv_initializer = Wqkv_initializer
+        self.Wo_initializer = Wo_initializer
+
+        self._reset_parameters()
+
+    def _reset_parameters(self) -> None:
+        for module in [self.Wq, self.Wk, self.Wv]:
+            create_initializer(self.Wqkv_initializer)(module.weight)
+            if module.bias is not None:
+                nn.init.zeros_(module.bias)
+
+        create_initializer(self.Wo_initializer)(self.Wo.weight)
+        if self.Wo.bias is not None:
+            nn.init.zeros_(self.Wo.bias)
+
+    @staticmethod
+    def split_heads(X_nqd: torch.Tensor, n_heads: int) -> torch.Tensor:
+        """Transposition for parallel computation of multiple attention heads."""
+        X_nqhk = X_nqd.reshape(X_nqd.shape[0], X_nqd.shape[1], n_heads, -1)
+        X_nhqk = X_nqhk.permute(0, 2, 1, 3)
+        return X_nhqk
+
+    @staticmethod
+    def merge_heads(X_nhqk: torch.Tensor) -> torch.Tensor:
+        """Reverse of split_heads."""
+        X_nqhk = X_nhqk.permute(0, 2, 1, 3)
+        X_nqd = X_nqhk.reshape(X_nqhk.shape[0], X_nqhk.shape[1], -1)
+        return X_nqd
+
+    def forward(
+        self,
+        query_ncd: torch.Tensor,
```

Comment:
`query_ncd` -> `x_query_ncd`
`key_ncd` -> `x_key_ncd`
`value_ncd` -> `x_value_ncd`

so that it is clear that these are the input embeddings.

Reply:
(just for the inpuit args, which only affected ln 125-127 -- the rest remains the same)